## Analitica Computacional para la toma de Decisiones
## Proyecto 1 - Bicicletas compartidas
Carlos Gomez - 202111593
<br>
Jeronimo Vargas - 202113305
<br>
Juan Esteban Lopez - 202021417
<br>

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn.model_selection as train_test_split
import sklearn.linear_model as linear_model

In [2]:
# Load the data
data = pd.read_csv('SeoulBikeDataClean.csv')

## Modelo de regresión 

$y = \beta_0 + \beta_1 \times \text{X1} + \beta_2 \times \text{X2} + \beta_3 \times \text{X3} + \beta_4 \times \text{X4} + \beta_5 \times \text{X5} + \beta_6 \times \text{X6} + \beta_7 \times \text{X7} + \beta_8 \times \text{X8} + \beta_9 \times \text{X9} + \beta_9 \times \text{X9} + \beta 10 \times \text{X10} + \beta 11 \times \text{X11} + \beta 12 \times \text{X12} + \beta 13 \times \text{X13} + \beta 14 \times \text{X14} $

In [3]:
indice = 0
for i in data.columns:
    if i != 'Rented Bike Count':
        print(f"X{indice}:", i)
    else:
        print("Y: ",i)
    indice += 1
    

Y:  Rented Bike Count
X1: Hour
X2: Temperature(C)
X3: Humidity(%)
X4: Wind speed (m/s)
X5: Visibility (10m)
X6: Dew point temperature(C)
X7: Solar Radiation (MJ/m2)
X8: Rainfall(mm)
X9: Snowfall (cm)
X10: Holiday
X11: Functioning Day
X12: Seasons_Spring
X13: Seasons_Summer
X14: Seasons_Winter


In [4]:
# Separar X y Y
X = data.drop('Rented Bike Count', axis=1)
Y = data['Rented Bike Count']
X.head()

Hour  Temperature(C)  Humidity(%)  Wind speed (m/s)  Visibility (10m)  \
0     0            -5.2           37               2.2              2000   
1     1            -5.5           38               0.8              2000   
2     2            -6.0           39               1.0              2000   
3     3            -6.2           40               0.9              2000   
4     4            -6.0           36               2.3              2000   

   Dew point temperature(C)  Solar Radiation (MJ/m2)  Rainfall(mm)  \
0                     -17.6                      0.0           0.0   
1                     -17.6                      0.0           0.0   
2                     -17.7                      0.0           0.0   
3                     -17.6                      0.0           0.0   
4                     -18.6                      0.0           0.0   

   Snowfall (cm)  Holiday  Functioning Day  Seasons_Spring  Seasons_Summer  \
0            0.0        0                1               0               0   
1            0.0        0                1               0               0   
2            0.0        0                1               0               0   
3            0.0        0                1               0               0   
4            0.0        0                1               0               0   

   Seasons_Winter  
0               1  
1               1  
2               1  
3               1  
4               1

In [5]:
Y.head()

0    254
1    204
2    173
3    107
4     78
Name: Rented Bike Count, dtype: int64

### División entre entrenamiento y prueba

In [6]:
import sklearn.model_selection as train_test_split

# Dividir los datos en entrenamiento y prueba
X_train, X_test, Y_train, Y_test = train_test_split.train_test_split(X, Y, test_size=0.2, random_state=0)

In [7]:
# Tamaño de los datos de entrenamiento y prueba
print("Tamaño de los datos de entrenamiento X:", X_train.shape)
print("Tamaño de los datos de prueba X:", X_test.shape)
print("Tamaño de los datos de entrenamiento Y:", Y_train.shape)
print("Tamaño de los datos de prueba Y:", Y_test.shape)

Tamaño de los datos de entrenamiento X: (7008, 14)
Tamaño de los datos de prueba X: (1752, 14)
Tamaño de los datos de entrenamiento Y: (7008,)
Tamaño de los datos de prueba Y: (1752,)


### Modelo RL

In [8]:
# Crear el modelo de regresión lineal
import sklearn.linear_model as lm
model = lm.LinearRegression()
model.fit(X_train, Y_train)

LinearRegression()

In [9]:
# Imprimir coeficientes
variables = X.columns
print("B0:", model.intercept_)
indice = 1
for i in model.coef_:
    feature = variables[indice-1]  
    print(f"B{indice} ({feature}):", i)
    indice += 1

B0: 53.98370399970281
B1 (Hour): 27.36755343612323
B2 (Temperature(C)): 14.43826753216662
B3 (Humidity(%)): -11.103888587537377
B4 (Wind speed (m/s)): 19.496121345722678
B5 (Visibility (10m)): 0.01745474269429792
B6 (Dew point temperature(C)): 12.820752006513155
B7 (Solar Radiation (MJ/m2)): -75.87692322007723
B8 (Rainfall(mm)): -55.83998938115196
B9 (Snowfall (cm)): 29.026841185932145
B10 (Holiday): -115.79557975811028
B11 (Functioning Day): 931.2135047213319
B12 (Seasons_Spring): -125.83055751023507
B13 (Seasons_Summer): -142.94948619323395
B14 (Seasons_Winter): -357.4897874909983


### Predicciones datos de prueba

In [10]:
y_pred = model.predict(X_test)

### Evaluar el modelo

**Error absoluto medio**: 

$$\text{MAE} = \frac 1n\sum_{i=1}^n|y_i-\hat{y}_i|$$

**Error cuadrado medio**: 
$$\text{MSE} = \frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2$$

**Raíz del Error cuadrado medio**: 
$$\text{RMSE} = \sqrt{\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2}$$

In [11]:
# Evaluar el modelo
import numpy as np
import sklearn.metrics as metrics

# Calcular el absoluto medio
print("MAE:", metrics.mean_absolute_error(Y_test, y_pred))

# Calcular el error cuadrático medio
print("MSE:", metrics.mean_squared_error(Y_test, y_pred))

# Raíz del error cuadrático medio
print("RMSE:", np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

# R cuadrado
print("R2:", metrics.r2_score(Y_test, y_pred))

MAE: 326.4871946346789
MSE: 190609.26508288088
RMSE: 436.58820996779207
R2: 0.5445663619093557


### Validacion Cruzada

In [12]:
# Validación cruzada
import sklearn.model_selection as model_selection
MAE_scores = model_selection.cross_val_score(model, X, Y, cv=5, scoring='neg_mean_absolute_error')
MSE_scores = model_selection.cross_val_score(model, X, Y, cv=5, scoring='neg_mean_squared_error')
RMSE_scores = model_selection.cross_val_score(model, X, Y, cv=5, scoring='neg_root_mean_squared_error')
R2_scores = model_selection.cross_val_score(model, X, Y, cv=5, scoring='r2')

# Imprimir los resultados
print("MAE:", -MAE_scores)
print("MAE promedio:", np.mean(-MAE_scores))

print("\nMSE:", -MSE_scores)
print("MSE promedio:", np.mean(-MSE_scores))

print("\nRMSE:", -RMSE_scores)
print("RMSE promedio:", np.mean(-RMSE_scores))

print("\nR2:", R2_scores)
print("R2 promedio:", np.mean(R2_scores))



MAE: [256.97307163 340.10509933 500.78085527 572.4185449  306.97031581]
MAE promedio: 395.4495773885684

MSE: [ 96040.02848479 171052.1881636  445813.25181357 438125.6601967
 168023.02306918]
MSE promedio: 263810.83034556854

RMSE: [309.90325665 413.58455987 667.69248297 661.91061345 409.90611494]
RMSE promedio: 492.59940557763946

R2: [-3.75979448  0.15871807  0.24172943 -0.08264446  0.55334   ]
R2 promedio: -0.5777302877899639


### Stastmodel

In [13]:
# Statsmodels
import sklearn as model_selection
import statsmodels.api as sm

# Añadir una constante
x = sm.add_constant(X)

# Ajustar el modelo
model = sm.OLS(Y, x).fit()

# Imprimir el resumen
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:      Rented Bike Count   R-squared:                       0.550
Model:                            OLS   Adj. R-squared:                  0.550
Method:                 Least Squares   F-statistic:                     764.8
Date:                Wed, 04 Sep 2024   Prob (F-statistic):               0.00
Time:                        14:45:48   Log-Likelihood:                -65598.
No. Observations:                8760   AIC:                         1.312e+05
Df Residuals:                    8745   BIC:                         1.313e+05
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   

### 1. Prueba de Significancia Global del Modelo
R-cuadrado (R-squared): El valor de R-squared es 0.550, lo que indica que aproximadamente el 55% de la variabilidad en la variable dependiente Rented Bike Count es explicada por las variables independientes en el modelo. Este es un nivel moderado de explicación, pero no excelente.
F-statistic y Prob(F-statistic): La F-statistic es 764.8 y su p-valor asociado es 0.000, lo que indica que el modelo en su conjunto es significativo. Esto significa que al menos una de las variables independientes tiene un efecto significativo en la variable dependiente.

### 2. Prueba de Significancia Individual de las Variables
Cada coeficiente estimado de las variables independientes tiene un t-statistic y un p-valor asociado que permiten evaluar la significancia de cada predictor individualmente:

#### Variables Significativas (p-valor < 0.05):

Hour (Hora): Muy significativa con un p-valor de 0.000. Indica que la hora del día tiene un impacto significativo en el número de bicicletas alquiladas.

Temperature(C) (Temperatura): Significativa con un p-valor de 0.000.

Humidity(%) (Humedad): Significativa con un p-valor de 0.000.

Wind speed (m/s) (Velocidad del Viento): Significativa con un p-valor de 0.000.

Dew point temperature(C) (Punto de rocío): Significativa con un p-valor de 0.004.

Solar Radiation (MJ/m2) (Radiación Solar): Muy significativa con un p-valor de 0.000.

Rainfall(mm) (Precipitación): Muy significativa con un p-valor de 0.000.

Snowfall (cm) (Nevadas): Significativa con un p-valor de 0.004.

Holiday (Día Festivo): Muy significativa con un p-valor de 0.000.

Functioning Day (Día de Funcionamiento): Muy significativa con un p-valor de 0.000.

Seasons_Spring, Seasons_Summer, Seasons_Winter (Temporada): Todas las variables de temporada son significativas con p-valores de 0.000.

#### Variable No Significativa (p-valor > 0.05):

Visibility (10m) (Visibilidad): Tiene un p-valor de 0.298, lo cual indica que no es estadísticamente significativa en el modelo. Esto sugiere que la visibilidad no tiene un efecto significativo en el número de bicicletas alquiladas.

### 3. Análisis de Test Estadísticos
Durbin-Watson Test: El valor de Durbin-Watson es 0.509, que indica la presencia de autocorrelación positiva en los residuos. Un valor cercano a 2 es ideal, y un valor tan bajo sugiere que los residuos no son independientes, lo cual puede afectar la validez del modelo.

Cond. No. (Condition Number): El número de condición es 3.27e+04, lo cual es bastante alto. Esto indica la posible presencia de multicolinealidad entre las variables independientes, lo que puede llevar a inestabilidades en los coeficientes del modelo.

### 4. Sugerencias para Mejorar el Modelo
Eliminar Variables No Significativas:

In [14]:
# Modelo con las variables más significativas
no_significativas = ["Visibility (10m)"]
X_new = X.drop(no_significativas, axis=1)
Y_new = data['Rented Bike Count']

# Dividir los datos en entrenamiento y prueba
X_train, X_test, Y_train, Y_test = train_test_split.train_test_split(X_new, Y_new, test_size=0.2, random_state=0)

# Crear el modelo de regresión lineal
model = lm.LinearRegression()
model.fit(X_train, Y_train)

# Imprimir coeficientes
variables = X_new.columns
print("B0:", model.intercept_)
indice = 1
for i in model.coef_:
    feature = variables[indice-1]  
    print(f"B{indice} ({feature}):", i)
    indice += 1

y_pred = model.predict(X_test)

# Evaluar el modelo
import numpy as np
import sklearn.metrics as metrics

# Calcular el absoluto medio
print("\nMAE:", metrics.mean_absolute_error(Y_test, y_pred))

# Calcular el error cuadrático medio
print("MSE:", metrics.mean_squared_error(Y_test, y_pred))

# Raíz del error cuadrático medio
print("RMSE:", np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

# R cuadrado
print("R2:", metrics.r2_score(Y_test, y_pred))

# Validación cruzada

MAE_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='neg_mean_absolute_error')

MSE_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='neg_mean_squared_error')

RMSE_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='neg_root_mean_squared_error')

R2_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='r2')

# Imprimir los resultados
print("MAE:", -MAE_scores)
print("MAE promedio:", np.mean(-MAE_scores))

print("\nMSE:", -MSE_scores)
print("MSE promedio:", np.mean(-MSE_scores))

print("\nRMSE:", -RMSE_scores)
print("RMSE promedio:", np.mean(-RMSE_scores))

print("\nR2:", R2_scores)
print("R2 promedio:", np.mean(R2_scores))

# Statsmodels
import sklearn as model_selection
import statsmodels.api as sm

# Añadir una constante
x = sm.add_constant(X_new)

# Ajustar el modelo
model = sm.OLS(Y_new, x).fit()

# Imprimir el resumen
print(model.summary())


B0: 114.91263270461275
B1 (Hour): 27.276154890634658
B2 (Temperature(C)): 14.031853760390954
B3 (Humidity(%)): -11.587298772370612
B4 (Wind speed (m/s)): 20.326906041477503
B5 (Dew point temperature(C)): 13.27830784804396
B6 (Solar Radiation (MJ/m2)): -78.58498045675061
B7 (Rainfall(mm)): -55.99701275657968
B8 (Snowfall (cm)): 28.76810589732168
B9 (Holiday): -115.19849981007042
B10 (Functioning Day): 930.4050566778026
B11 (Seasons_Spring): -131.21150862989555
B12 (Seasons_Summer): -141.40168452043733
B13 (Seasons_Winter): -363.30009575336754

MAE: 326.60015025036734
MSE: 190408.074931489
RMSE: 436.357737334276
R2: 0.5450470770653402
MAE: [257.04284118 340.36921151 499.6997579  559.61381718 307.65526172]
MAE promedio: 392.87617789615274

MSE: [ 96135.69858181 171153.85113503 443854.11015196 420726.42070605
 168146.32781126]
MSE promedio: 260003.2816772231

RMSE: [310.05757301 413.70744631 666.22376883 648.63427346 410.05649344]
RMSE promedio: 489.73591101024766

R2: [-3.76453594  0.1582

### Sugerencias para Mejorar el Modelo

Verificar y Manejar la Multicolinealidad:

Utiliza el VIF (Variance Inflation Factor) para identificar y tratar la multicolinealidad. Si alguna de las variables tiene un VIF alto (mayor a 10), podrías considerar eliminarla o combinarla con otras variables correlacionadas.

In [15]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Calcular VIF para cada variable independiente
X = data[['Hour', 'Temperature(C)', 'Humidity(%)', 'Wind speed (m/s)', 'Dew point temperature(C)',
          'Solar Radiation (MJ/m2)', 'Rainfall(mm)', 'Snowfall (cm)', 'Holiday', 'Functioning Day',
          'Seasons_Spring', 'Seasons_Summer', 'Seasons_Winter']]

vif_data = pd.DataFrame()
vif_data["Variable"] = X.columns
vif_data["VIF"] = [variance_inflation_factor(X.values, i) for i in range(len(X.columns))]

print(vif_data)

                    Variable        VIF
0                       Hour   4.517189
1             Temperature(C)  32.341156
2                Humidity(%)  19.117206
3           Wind speed (m/s)   4.854020
4   Dew point temperature(C)  20.408749
5    Solar Radiation (MJ/m2)   2.745663
6               Rainfall(mm)   1.076088
7              Snowfall (cm)   1.139975
8                    Holiday   1.075511
9            Functioning Day  30.137552
10            Seasons_Spring   2.155776
11            Seasons_Summer   3.481466
12            Seasons_Winter   4.303814


### Modelo revisando la multicolinealidad con el VIF de las variables. 

In [16]:
# Eliminar las variables con VIF mayor a 10
variables_eliminar = ['Temperature(C)', 'Humidity(%)', 'Dew point temperature(C)']
X_new = X.drop(variables_eliminar, axis=1)
Y_new = data['Rented Bike Count']

# Dividir los datos en entrenamiento y prueba
X_train, X_test, Y_train, Y_test = train_test_split.train_test_split(X_new, Y_new, test_size=0.2, random_state=0)

# Crear el modelo de regresión lineal
model = lm.LinearRegression()
model.fit(X_train, Y_train)

# Imprimir coeficientes
variables = X_new.columns
print("B0:", model.intercept_)
indice = 1
for i in model.coef_:
    feature = variables[indice-1]  
    print(f"B{indice} ({feature}):", i)
    indice += 1

y_pred = model.predict(X_test)

# Evaluar el modelo
import numpy as np
import sklearn.metrics as metrics

# Calcular el absoluto medio
print("\nMAE:", metrics.mean_absolute_error(Y_test, y_pred))

# Calcular el error cuadrático medio
print("MSE:", metrics.mean_squared_error(Y_test, y_pred))

# Raíz del error cuadrático medio
print("RMSE:", np.sqrt(metrics.mean_squared_error(Y_test, y_pred)))

# R cuadrado
print("R2:", metrics.r2_score(Y_test, y_pred))

# Validación cruzada

MAE_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='neg_mean_absolute_error')

MSE_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='neg_mean_squared_error')

RMSE_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='neg_root_mean_squared_error')

R2_scores = model_selection.model_selection.cross_val_score(model, X_new, Y_new, cv=5, scoring='r2')

# Imprimir los resultados
print("MAE:", -MAE_scores)
print("MAE promedio:", np.mean(-MAE_scores))

print("\nMSE:", -MSE_scores)
print("MSE promedio:", np.mean(-MSE_scores))

print("\nRMSE:", -RMSE_scores)
print("RMSE promedio:", np.mean(-RMSE_scores))

print("\nR2:", R2_scores)
print("R2 promedio:", np.mean(R2_scores))

# Statsmodels
import sklearn as model_selection
import statsmodels.api as sm

# Añadir una constante
x = sm.add_constant(X_new)

# Ajustar el modelo
model = sm.OLS(Y_new, x).fit()

# Imprimir el resumen
print(model.summary())



B0: -433.66429241486844
B1 (Hour): 35.336824225536795
B2 (Wind speed (m/s)): 26.94864243602467
B3 (Solar Radiation (MJ/m2)): 74.97739091790991
B4 (Rainfall(mm)): -79.96931779367085
B5 (Snowfall (cm)): -60.07248237726964
B6 (Holiday): -62.36934985724537
B7 (Functioning Day): 878.6447136625934
B8 (Seasons_Spring): -185.14021791406438
B9 (Seasons_Summer): 112.01384899844486
B10 (Seasons_Winter): -671.7232136429602

MAE: 353.31449100406513
MSE: 220846.78769556028
RMSE: 469.9433877559725
R2: 0.47231811666087475
MAE: [247.28513029 430.82489374 521.5967792  491.3875826  419.43191668]
MAE promedio: 422.10526050201133

MSE: [ 89453.64517239 270472.31574112 491949.42155096 349508.60466423
 253980.17613031]
MSE promedio: 291072.8326518004

RMSE: [299.08802245 520.06952972 701.39106178 591.19252758 503.96445919]
RMSE promedio: 523.1411201445992

R2: [-3.43336985 -0.33025758  0.16325779  0.13633555  0.32483784]
R2 promedio: -0.6278392493896108
                            OLS Regression Results     

Para analizar los tres modelos (M1, M2, y M3) y determinar cuál es el mejor, vamos a comparar sus métricas de rendimiento, como MAE (Mean Absolute Error), MSE (Mean Squared Error), RMSE (Root Mean Squared Error), y R² (R-squared), además de interpretar los resultados de las regresiones OLS para cada uno de los modelos.

### Análisis de los Modelos
#### Modelo 1 (M1):

MAE Promedio: 395.45

MSE Promedio: 263,810.83

RMSE Promedio: 492.60

R² Promedio: -0.58 (R² de cada pliegue varía significativamente, algunos negativos)


OLS Regression Results:
R²: 0.550

Cond. No.: 3.27e+04 (indica posible multicolinealidad)

Interpretación:

M1 tiene un MAE promedio y RMSE relativamente altos, lo que indica un error considerable en las predicciones.
El R² de -0.58 en promedio y valores negativos en algunos pliegues sugieren que el modelo no generaliza bien en todos los pliegues.
El valor de R² global de 0.550 muestra que el modelo explica el 55% de la variabilidad en los datos, pero la multicolinealidad es un problema debido a un alto número de condición.

#### Modelo 2 (M2):

MAE Promedio: 392.88

MSE Promedio: 260,003.28

RMSE Promedio: 489.74

R² Promedio: -0.57 (mejor que M1, pero todavía hay valores negativos)

OLS Regression Results:
R²: 0.550

Cond. No.: 1.26e+03 (mucho menor que M1)

Interpretación:

M2 tiene una ligera mejora en MAE, MSE, y RMSE en comparación con M1.
El R² global es el mismo que en M1, pero el número de condición es mucho más bajo, lo que indica menos multicolinealidad en el modelo.
Las métricas mejoran ligeramente, pero no significativamente en comparación con M1. A pesar de estas mejoras, el modelo todavía muestra problemas con generalización en algunos pliegues.

#### Modelo 3 (M3):

MAE Promedio: 422.11

MSE Promedio: 291,072.83

RMSE Promedio: 523.14

R² Promedio: -0.63 (peor desempeño en comparación con M1 y M2)

OLS Regression Results:
R²: 0.475

Cond. No.: 108 (muy bajo, indica muy poca o ninguna multicolinealidad)

Interpretación:

M3 tiene los peores valores de MAE, MSE y RMSE entre los tres modelos, lo que indica errores más grandes en las predicciones.
El R² global es de 0.475, el más bajo entre los tres modelos, lo que indica que este modelo explica menos la variabilidad de los datos.
A pesar de tener el número de condición más bajo, lo que indica menos multicolinealidad, los errores son más grandes y el R² promedio es el más bajo.
Conclusión

### Comparación General de los Modelos:

Rendimiento Predictivo (MAE, MSE, RMSE):

M2 tiene el mejor rendimiento predictivo, con el menor MAE, MSE y RMSE en promedio. Esto sugiere que M2 es el modelo con los errores más bajos en general.
Coeficiente de Determinación (R²):

Aunque M1 y M2 tienen el mismo R² global (0.550), M2 tiene un promedio de R² ligeramente mejor (-0.57) en comparación con M1 (-0.58), lo que sugiere que generaliza mejor.
Multicolinealidad:

M2 presenta un número de condición significativamente más bajo que M1 (1.26e+03 vs. 3.27e+04), lo que indica menos problemas de multicolinealidad, lo cual es deseable.
M3 tiene el número de condición más bajo (108), pero su rendimiento predictivo es el peor, lo que indica que la eliminación de algunas variables clave no mejoró el modelo.
Mejor Modelo
El Mejor Modelo es el Modelo 2 (M2) debido a su mejor rendimiento predictivo (menores MAE, MSE, RMSE), un mejor R² promedio y menor multicolinealidad en comparación con M1. Aunque M3 tiene un número de condición bajo, su rendimiento predictivo es inferior al de M1 y M2.